src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs

```diff
@@ -93,41 +93,77 @@ internal partial class PackageInstallerService : AbstractDelayStartedService, IP
+            lock (_gate)
             {
-                return packageSources;
+                if (_packageSourcesAsync is null)
+                {
+                    _packageSourcesAsync = ThreadingContext.JoinableTaskFactory.RunAsync(() => GetPackageSourcesImplAsync());
+                }
+
+                packageSourcesAsync = _packageSourcesAsync;
             }
 
-            try
+            if (packageSourcesAsync.IsCompleted)
             {
-                packageSources = _packageSourceProvider.Value.GetSources(includeUnOfficial: true, includeDisabled: false)
-                    .SelectAsArray(r => new PackageSource(r.Key, r.Value));
+                // Since the task is already completed, we know this 'await' will complete synchronously.
+                return await packageSourcesAsync;
             }
-            catch (Exception ex) when (ex is InvalidDataException || ex is InvalidOperationException)
+            else if (allowSwitchToMainThread)
             {
-                // These exceptions can happen when the nuget.config file is broken.
-                packageSources = ImmutableArray<PackageSource>.Empty;
+                return await packageSourcesAsync.JoinAsync(cancellationToken).ConfigureAwait(false);
             }
-            catch (ArgumentException ae) when (FatalError.ReportWithoutCrash(ae))
+            else
             {
-                // This exception can happen when the nuget.config file is broken, e.g. invalid credentials.
-                // https://github.com/dotnet/roslyn/issues/40857
-                packageSources = ImmutableArray<PackageSource>.Empty;
+                // The result was not available and switching to the main thread is not allowed. Return without caching
+                // a result.
+                return null;
             }
+        }
 
-            var previousPackageSources = ImmutableInterlocked.InterlockedCompareExchange(ref _packageSources, packageSources, default);
-            if (previousPackageSources != null)
+        private async Task<ImmutableArray<PackageSource>> GetPackageSourcesImplAsync()
+        {
+            // Automatically retry the operation if it gets cancelled due to a change in the workspace.
+            while (true)
             {
-                // Another thread already initialized _packageSources
-                packageSources = previousPackageSources;
+                var cancellationToken = _tokenSource.Token;
+                try
+                {
+                    return await GetPackageSourcesImplAsync(cancellationToken).ConfigureAwait(false);
+                }
+                catch (OperationCanceledException) when (cancellationToken.IsCancellationRequested)
+                {
+                    // The failure may have been caused by a workspace change; try again after a short delay
+                    await Task.Delay(TimeSpan.FromMilliseconds(250)).ConfigureAwait(false);
+                }
             }
```

Comment:
i don't think we need cancellation/retry here at all.

Reply:
➡️ This discussion is covered by https://github.com/dotnet/roslyn/pull/43718#discussion_r424001591